--- a/cmd/mvebu/Kconfig	2019-01-31 10:34:20.000000000 +0300
+++ b/cmd/mvebu/Kconfig	2019-02-01 09:58:13.000000000 +0300
@@ -9,6 +9,12 @@
 	  For details about bubt command please see the documentation
 	  in doc/mvebu/cmd/bubt.txt
 
+config CMD_MVEBU_SDFLASH
+	bool "sdflash"
+	default n
+	help
+	  sdflash - Burn a u-boot image from SD-card to eMMC storage
+
 config CMD_MVEBU_PHY_FW_DOWNLOAD
 	bool "phy_fw_down"
 	depends on MVMDIO
